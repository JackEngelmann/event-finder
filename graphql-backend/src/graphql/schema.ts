import { gql } from 'apollo-server-express'

export const typeDefs = gql`
    type Club {
        id: Int!
        name: String!
    }

    type Event {
        id: Int!
        name: String!
        description: String
        date: String!
        club: Club!
        imageUrl: String
    }

    type Query {
        club(id: Int!): Club
        clubs: [Club!]!
        event(id: Int!): Event
        events: [Event!]!
    }
`
