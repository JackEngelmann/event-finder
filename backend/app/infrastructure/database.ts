import 'reflect-metadata'
import { databaseConfig } from '../../databaseConfig'
import { createConnection, Connection } from 'typeorm'
import { ClubDataModel } from '../components/club/orm/club'
import { EventDataModel } from '../components/event/orm/event'
import { EventGenreDataModel } from '../components/genre/orm/eventGenre'
import { GenreDataModel } from '../components/genre/orm/genre'
import { ImageDataModel } from '../components/image/orm/image'
import { UserDataModel } from '../components/auth/orm/user'
import { AppliedScriptDataModel } from '../components/scripts/orm/appliedScripts'
import { EventImageDataModel } from '../components/image/orm/eventImage'
import { ClubImageDataModel } from '../components/image/orm/clubImage'

export const createDbConnection = (dbName?: any) => {
    return createConnection({
        name: dbName,
        logging: ['query', 'error', 'info', 'log', 'warn'],
        entities: [
            AppliedScriptDataModel,
            ClubDataModel,
            EventDataModel,
            EventGenreDataModel,
            GenreDataModel,
            ImageDataModel,
            UserDataModel,
            EventImageDataModel,
            ClubImageDataModel,
        ],
        ...databaseConfig.connectionOptions,
        logger: 'file',
    })
}

const connectionPromise = createDbConnection()

export async function getDbConnection() {
    return await connectionPromise
}
