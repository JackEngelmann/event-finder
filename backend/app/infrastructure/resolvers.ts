import { IResolvers } from 'graphql-tools'
import { queryEvent } from '../components/event/queries/event'
import { queryEvents } from '../components/event/queries/events'
import { queryClubs } from '../components/club/queries/clubs'
import { queryClub } from '../components/club/queries/club'
import { queryGenre } from '../components/genre/queries/genre'
import { queryGenres } from '../components/genre/queries/genres'
import { queryGenresForEvent } from '../components/genre/queries/genresForEvent'
import { createEvent } from '../components/event/commands/createEvent'
import { updateEvent } from '../components/event/commands/updateEvent'
import { deleteEvent } from '../components/event/commands/deleteEvent'
import { createClub } from '../components/club/commands/createClub'
import { updateClub } from '../components/club/commands/updateClub'
import { deleteClub } from '../components/club/commands/deleteClub'
import { AppContext } from './appContext'
import { requireAdminPermission } from '../components/auth/authentication'
import { Logger } from './logger'
import { uploadImage } from '../components/image/commands/uploadImages'
import { queryImageUrlsForEvent } from '../components/image/queries/imageUrlsForEvent'
import { queryImageUrlsForClub } from '../components/image/queries/imageUrlsForClub'
import { queryEventsForClub } from '../components/event/queries/eventsForClub'
import { queryLinksForEvent } from '../components/link/queries/linksForEvent'
import { queryLinksForClub } from '../components/link/queries/linksForClub'

type Source = {
    clubId: number
    id: number
}

const logger = new Logger()

export const resolvers: IResolvers<Source, AppContext> = {
    Query: {
        clubs: (obj, args, appContext, info) => {
            logger.info('query clubs')
            return queryClubs(appContext)
        },
        club: (obj, args, appContext, info) => {
            logger.info('query club')
            return queryClub(appContext, args.id)
        },
        events: (obj, args, appContext, info) => {
            logger.info('query events')
            return queryEvents(appContext, args.filter)
        },
        event: (obj, args, appContext, info) => {
            logger.info('query event')
            return queryEvent(appContext, args.id)
        },
        genre: (obj, args, appContext, info) => {
            logger.info('query genre')
            return queryGenre(appContext, args.id)
        },
        genres: (obj, args, appContext, info) => {
            logger.info('query genres')
            return queryGenres(appContext)
        },
        me: (obj, args, appContext, info) => {
            logger.info('query me')
            return appContext.user
        },
    },
    Mutation: {
        createEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            const eventId = await createEvent(appContext, args.input)
            const event = await queryEvent(appContext, eventId)
            return { event }
        },
        updateEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await updateEvent(appContext, args.input)
            const event = await queryEvent(appContext, args.input.id)
            return { event }
        },
        deleteEvent: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await deleteEvent(appContext, args.id)
            return { id: args.id }
        },
        createClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            const clubId = await createClub(appContext, args.input)
            const club = await queryClub(appContext, clubId)
            return { club }
        },
        updateClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await updateClub(appContext, args.input)
            const club = await queryClub(appContext, args.input.id)
            return { club }
        },
        deleteClub: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            await deleteClub(appContext, args.id)
            return { id: args.id }
        },
        uploadImage: async (obj, args, appContext, info) => {
            requireAdminPermission(appContext)
            const imageUrl = await uploadImage(appContext, args.input)
            return { imageUrl }
        },
    },
    Event: {
        club: (event, args, appContext, info) => {
            return queryClub(appContext, event.clubId)
        },
        genres: (event, args, appContext, info) =>
            queryGenresForEvent(appContext, event.id),
        imageUrls: (event, args, appContext, info) =>
            queryImageUrlsForEvent(appContext, event.id),
        links: (event, args, appContext, info) =>
            queryLinksForEvent(appContext, event.id),
    },
    Club: {
        imageUrls: (club, args, appContext, info) =>
            queryImageUrlsForClub(appContext, club.id),
        events: (club, args, appContext, info) =>
            queryEventsForClub(appContext, club.id, args.fromDay),
        links: (club, args, appContext, info) =>
            queryLinksForClub(appContext, club.id),
    },
}
