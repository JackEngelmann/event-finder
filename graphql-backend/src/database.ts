import sqlite3 from 'sqlite3'
import { insertClubsIfNoneExist } from './seed-data/clubData'
import { insertEventsIfNoneExist } from './seed-data/eventData'

sqlite3.verbose()

const DB_SOURCE = 'db.sqlite'
export const db = new sqlite3.Database(DB_SOURCE, async err => {
    if (err) {
        console.error(err.message)
        throw err
    }
    db.serialize(() => {
        createTables()
        insertClubsIfNoneExist(db)
        insertEventsIfNoneExist(db)
    })
})

async function createTables() {
    db.run(`
        CREATE TABLE IF NOT EXISTS club (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text
        )
    `)
    db.run(`
        CREATE TABLE IF NOT EXISTS event (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text,
            description text,
            date string,
            clubId integer,
            imageUrl text
        )
    `)
}
