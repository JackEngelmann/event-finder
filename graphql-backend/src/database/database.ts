import sqlite3, { Database } from 'sqlite3'
import { createSchema } from './createSchema'
import { applySeeds } from './applySeeds'

sqlite3.verbose()

const DB_SOURCE = 'db.sqlite'

export const db = new sqlite3.Database(DB_SOURCE, async err => {
    if (err) {
        console.error(err.message)
        throw err
    }
    db.serialize(() => {
        createSchema(db)
        applySeeds(db)
    })
})

export const createTestDb = async() => new Promise<Database>(resolve => {
    const db = new sqlite3.Database(':memory:', async err => {
        if (err) {
            console.error(err.message)
            throw err
        }
    })
    db.serialize(async () => {
        createSchema(db)
        resolve(db)
    })
})
