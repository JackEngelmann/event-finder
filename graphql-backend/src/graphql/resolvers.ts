import { IResolvers } from "graphql-tools";
import { Database } from "sqlite3";
import { queryEvent } from "../queries/event";
import { queryEvents } from "../queries/events";
import { queryClubs } from "../queries/clubs";
import { queryClub } from "../queries/club";
import { queryGenre } from "../queries/genre";
import { queryGenres } from "../queries/genres";
import { queryGenresForEvent } from "../queries/genresForEvent";
import { createEvent } from "../mutations/createEvent";

type Source = {
    clubId: number
    id: number
}
type Context = {
    db: Database
}

export const resolvers: IResolvers<Source, Context> = {
    Query: {
        clubs: (obj, args, appContext, info) => queryClubs(appContext),
        club: (obj, args, appContext, info) => queryClub(appContext, parseInt(args.id)),
        events: (obj, args, appContext, info) => queryEvents(appContext, args.filter),
        event: (obj, args, appContext, info) => queryEvent(appContext, parseInt(args.id)),
        genre: (obj, args, appContext, info) => queryGenre(appContext, parseInt(args.id)),
        genres: (obj, args, appContext, info) => queryGenres(appContext),
    },
    Mutation: {
        createEvent: async (obj, args, appContext, info) => {
            const eventId = await createEvent(appContext, args.input)
            const event = await queryEvent(appContext, eventId)
            return { event }
        },
    },
    Event: {
        club: (event, args, appContext, info) => queryClub(appContext, event.clubId),
        genres: (event, args, appContext, info) => queryGenresForEvent(appContext, event.id)
    }
}