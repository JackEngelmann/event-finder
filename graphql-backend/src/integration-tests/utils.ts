import { Database } from "sqlite3"

export const genreFragment = `
    id
    name
`

export const eventFragment = `
    club {
        id
        name
    }
    name,
    description,
    date,
    imageUrl,
    priceCategory,
    admissionFee,
    admissionFeeWithDiscount,
    special,
    minimumAge,
    amountOfFloors
    genres {
        ${genreFragment}
    }
`

export const clubFragment = `
    name
    address
    region
    contact
    email
    specials
    description
    link
`

export function insertTestData(db: Database) {
    return new Promise(resolve => {
        db.serialize(() => {
            db.run(`
                    INSERT INTO club (
                        [name],
                        [address],
                        [region],
                        [contact],
                        [email],
                        [specials],
                        [description],
                        [link]
                    ) VALUES (
                        'name',
                        'address',
                        'region',
                        'contact',
                        'email',
                        'specials',
                        'description',
                        'link'
                    )
                `)
            db.run(`
                    INSERT INTO club (
                        [name],
                        [address],
                        [region],
                        [contact],
                        [email],
                        [specials],
                        [description],
                        [link]
                    ) VALUES (
                        'name',
                        'address',
                        'region',
                        'contact',
                        'email',
                        'specials',
                        'description',
                        'link'
                    )
                `)
            db.run(`
                    INSERT INTO event
                    (
                        name,
                        description,
                        date,
                        clubId,
                        imageUrl,
                        priceCategory,
                        admissionFee,
                        admissionFeeWithDiscount,
                        special,
                        minimumAge,
                        amountOfFloors
                    ) 
                    VALUES (
                        'name',
                        'description',
                        'date',
                        1,
                        'imageUrl',
                        2,
                        10.0,
                        20.0,
                        'specials',
                        18,
                        4
                    )
                `)
            db.run(`
                    INSERT INTO genre ([name]) VALUES
                        ('genre1'),
                        ('genre2'),
                        ('genre3')
                `)
            db.run(
                `
                    INSERT INTO eventGenre (eventId, genreId) VALUES (1, 1)
                `,
                () => resolve()
            )
        })
    })
}
