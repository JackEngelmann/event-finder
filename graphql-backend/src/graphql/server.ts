import { ApolloServer } from 'apollo-server-express'
import { typeDefs } from './schema'
import { db } from '../database'
import { resolvers } from './resolvers'
import express from 'express'
import http, { Server } from 'http'

const PORT = 5000

const app = express()

app.use(express.static('public'))

const apolloServer = new ApolloServer({
    typeDefs,
    resolvers,
    context: { db },
    playground: true,
    introspection: true
})

apolloServer.applyMiddleware({ app })

let expressServer: Server | null = null

export async function startServer() {
    expressServer = http.createServer(app).listen({ port: PORT }, () => {
        console.log('started')
    })
}

export async function stopServer() {
    if (expressServer) expressServer.close()
}


