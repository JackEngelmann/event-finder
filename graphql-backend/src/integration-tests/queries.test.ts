import { createTestClient } from 'apollo-server-testing'
import { ApolloServer } from 'apollo-server-express'
import { typeDefs } from '../graphql/schema'
import { resolvers } from '../graphql/resolvers'
import { createTestDb } from '../database/database'
import { genreFragment, eventFragment, clubFragment, insertTestData } from './utils'

const server = new ApolloServer({
    typeDefs,
    resolvers,
    context: async () => {
        const db = await createTestDb()
        await insertTestData(db)
        return { db }
    },
})

const { query } = createTestClient(server)

describe('should query', () => {
    test('clubs correctly', async done => {
        const result = await query({
            query: `{
                clubs { ${clubFragment} }
            }`,
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
        done()
    })
    test('events correctly', async done => {
        const result = await query({
            query: `{
                events { ${eventFragment} }
            }`,
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
        done()
    })
    test('genres correctly', async done => {
        const result = await query({
            query: `
            {
                genres { ${genreFragment} }
            }
            `,
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
        done()
    })
    test('existing club correctly', async done => {
        const result = await query({
            query: `
                query clubById($id: Int!) {
                    club(id: $id) { ${clubFragment} }
                }
            `,
            variables: {
                id: 1
            }
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
        done()
    })
    test('existing event correctly', async done => {
        const result = await query({
            query: `
                query eventQuery($id: Int!) {
                    event(id: $id) { ${eventFragment} }
                }
            `,
            variables: {
                id: 1
            }
        })
        expect(result.data).toBeDefined()
        expect(result).toMatchSnapshot()
        done()
    })
    test('not existing event and return undefined', async done => {
        const result = await query({
            query: `
                query eventQuery($id: Int!) {
                    event(id: $id) { ${eventFragment} }
                }
            `,
            variables: {
                id: 2
            }
        })
        expect(result.data!.event).toBeNull()
        done()
    })
})
