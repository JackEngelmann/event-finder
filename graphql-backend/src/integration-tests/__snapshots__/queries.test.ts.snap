// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should query clubs correctly 1`] = `
Object {
  "data": Object {
    "clubs": Array [
      Object {
        "address": "address",
        "contact": "contact",
        "description": "description",
        "email": "email",
        "link": "link",
        "name": "name",
        "region": "region",
        "specials": "specials",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`should query events correctly 1`] = `
Object {
  "data": Object {
    "events": Array [
      Object {
        "admissionFee": 10,
        "admissionFeeWithDiscount": 20,
        "amountOfFloors": 4,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "date",
        "description": "description",
        "genres": Array [
          Object {
            "id": 1,
            "name": "genre1",
          },
        ],
        "imageUrl": "imageUrl",
        "minimumAge": 18,
        "name": "name",
        "priceCategory": 2,
        "special": "specials",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`should query existing club correctly 1`] = `
Object {
  "data": Object {
    "club": Object {
      "address": "address",
      "contact": "contact",
      "description": "description",
      "email": "email",
      "link": "link",
      "name": "name",
      "region": "region",
      "specials": "specials",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`should query existing event correctly 1`] = `
Object {
  "data": Object {
    "event": Object {
      "admissionFee": 10,
      "admissionFeeWithDiscount": 20,
      "amountOfFloors": 4,
      "club": Object {
        "id": 1,
        "name": "name",
      },
      "date": "date",
      "description": "description",
      "genres": Array [
        Object {
          "id": 1,
          "name": "genre1",
        },
      ],
      "imageUrl": "imageUrl",
      "minimumAge": 18,
      "name": "name",
      "priceCategory": 2,
      "special": "specials",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`should query genres correctly 1`] = `
Object {
  "data": Object {
    "genres": Array [
      Object {
        "id": 1,
        "name": "genre1",
      },
      Object {
        "id": 2,
        "name": "genre2",
      },
      Object {
        "id": 3,
        "name": "genre3",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
